core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -366,25 +390,72 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
                 // A task should only be in the aborted state if the STAGE is done (ABORTED or FAILED)
                 fail(new TrinoException(GENERIC_INTERNAL_ERROR, "A task is in the ABORTED state but stage is " + stageState));
                 break;
-            case FLUSHING:
-                flushingTasks.add(taskStatus.getTaskId());
-                break;
-            case FINISHED:
-                finishedTasks.add(taskStatus.getTaskId());
```

Comment:
How this is now handled?

It seems like:
```
finishedTasks.add(taskStatus.getTaskId());
flushingTasks.remove(taskStatus.getTaskId());
```
should really be an atomic change.

Reply:
Now, we are counting the number of `finished` tasks (when all tasks are finished it means that the stage is finished)  and number of `flushing` or `finished` tasks.
When all tasks are `finished` or `flushing` and there was at least one `flushing` task the stage can be considered as `FLUSHING`.